job.go

```
@@ -0,0 +1,236 @@
+		Error     error // go error
+	}
+
+	Job struct {
+		// Unique Job Id
+		Id string
+
+		// Set underlying os/exec.Cmd.Path
+		Path string
+
+		// Set underlying os/exec.Cmd.Args.
+		Args []string
+
+		// $MAJ:$MIN device number for MNT namespace
+		Mnt string
+
+		// CPUPercent represents the quota of all cores.
+		CPUPercent int
+
+		// MemoryMB represents the quota of memory to in Megabytes.
+		MemoryMB int
+
+		// ReadBPS represents the max bytes read per second by proc
+		ReadBPS int
+
+		// WriteBPS represents the max bytes write per second by proc
+		WriteBPS int
+
+		// Read-write lock for job state
+		rw sync.RWMutex
+
+		// jail path to isolate process
+		jailPath string
+
+		// log file bind to os/exec.Cmd.Stdout and os/exec.Cmd.Stderr
+		logs *os.File
+
+		// cgroup file assigned to job
+		cgroup *os.File
+
+		status Status
+
+		doneCh chan struct{} // closed when done running
+	}
+
+	JobReader struct {
+		job     *Job
+		logs    io.ReadCloser
+		inotify *os.File
+	}
+)
+
+func Init() error {
+	// allow starting job
+	startable = true
+
+	if len(os.Args) < 3 && os.Args[1] == JailOp {
+		return ErrInvalidArgs
+	}
+	// skip run command in jail
+	if len(os.Args) < 3 || os.Args[1] != JailOp {
+		return nil
+	}
+	fmt.Println(os.Args)
+	// disallow starting job on jailed proc
+	startable = false
+
+	if err := mount(); err != nil {
+		return err
+	}
+	args := os.Args[2:]
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+	return fmt.Errorf(cmd.ProcessState.String())
+}
+
+// Started return true if StartedAt is not zero
+func (s Status) Started() bool {
+	return !s.StartedAt.IsZero()
+}
+
+// Stopped return true if StoppedAt is not zero
+func (s Status) Stopped() bool {
+	return !s.StoppedAt.IsZero()
+}
+
+// Start starts the command
+func (j *Job) Start(ctx context.Context) error {
+	if !startable {
+		panic("unsafe start")
```

panicking isn't very cool, could you return an error instead?

---

good point